product/scanning.py

```diff
@@ -108,40 +108,110 @@ def retrieve_scan_result(
         product = self._api.scan_result(product_id=product_id)
         return product
 
-    def upload_image(
-            self,
-            container_image: ContainerImage,
-            component: Component,
-            wait_for_result: bool=False
-        ):
-        metadata = self._metadata(container_image=container_image, component=component)
+    def process_image(
+        self,
+        container_image: ContainerImage,
+        component: Component,
+        processing_mode: ProcessingMode,
+        wait_for_result: bool=False,
+    ):
+        # Fetch latest known analysis result and store for processing
+        self._last_analysis_result = self.retrieve_scan_result(
+            container_image= container_image,
+            component=component,
+        )
+
+        if processing_mode is ProcessingMode.UPLOAD_IF_CHANGED:
+            processing_result = self.process_upload(
+                container_image=container_image,
+                component=component,
+            )
+        else:
+            raise NotImplementedError
+
+        if processing_result.status is UploadStatus.SKIPPED:
+            return processing_result
+
+        if wait_for_result:
+            processing_result = self.wait_for_result(processing_result)
+
+        return self.finalize_result(processing_result)
+
+    def finalize_result(self, upload_result: UploadResult):
```

Comment:
after removal of optional waiting for result, the entire method should be no longer needed